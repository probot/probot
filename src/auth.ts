import type { Logger } from "pino";

import { getAuthenticatedOctokit } from "./octokit/get-authenticated-octokit";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { State } from "./types";

/**
 * Authenticate and get a GitHub client that can be used to make API calls.
 *
 * You'll probably want to use `context.octokit` instead.
 *
 * **Note**: `app.auth` is asynchronous, so it needs to be prefixed with a
 * [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)
 * to wait for the magic to happen.
 *
 * ```js
 *  module.exports = (app) => {
 *    app.on('issues.opened', async context => {
 *      const octokit = await app.auth();
 *    });
 *  };
 * ```
 *
 * @param id - ID of the installation, which can be extracted from
 * `context.payload.installation.id`. If called without this parameter, the
 * client wil authenticate [as the app](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app)
 * instead of as a specific installation, which means it can only be used for
 * [app APIs](https://developer.github.com/v3/apps/).
 *
 * @returns An authenticated GitHub API client
 */
export async function auth(
  state: State,
  installationId?: number,
  log?: Logger
): Promise<InstanceType<typeof ProbotOctokit>> {
  return getAuthenticatedOctokit(
    Object.assign({}, state, log ? { log } : null),
    installationId
  );
}
