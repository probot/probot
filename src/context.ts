import path from "path";

import { EmitterWebhookEvent as WebhookEvent } from "@octokit/webhooks";
import merge from "deepmerge";

import type { Logger } from "pino";

import { ProbotOctokit } from "./octokit/probot-octokit";
import { aliasLog } from "./helpers/alias-log";
import { DeprecatedLogger } from "./types";
import { EmitterWebhookEventName as WebhookEvents } from "@octokit/webhooks/dist-types/types";

export type MergeOptions = merge.Options;

/**
 * The context of the event that was triggered, including the payload and
 * helpers for extracting information can be passed to GitHub API calls.
 *
 *  ```js
 *  module.exports = app => {
 *    app.on('push', context => {
 *      context.log.info('Code was pushed to the repo, what should we do with it?');
 *    });
 *  };
 *  ```
 *
 * @property {octokit} octokit - An Octokit instance
 * @property {payload} payload - The webhook event payload
 * @property {log} log - A pino instance
 */
export class Context<E extends WebhookEvents = WebhookEvents> {
  public name: WebhookEvents;
  public id: string;
  public payload: WebhookEvent<E>["payload"];

  public octokit: InstanceType<typeof ProbotOctokit>;
  public log: DeprecatedLogger;

  constructor(
    event: WebhookEvent<E>,
    octokit: InstanceType<typeof ProbotOctokit>,
    log: Logger
  ) {
    this.name = event.name;
    this.id = event.id;
    this.payload = event.payload;

    this.octokit = octokit;
    this.log = aliasLog(log);
  }

  /**
   * Return the `owner` and `repo` params for making API requests against a
   * repository.
   *
   * ```js
   * const params = context.repo({path: '.github/config.yml'})
   * // Returns: {owner: 'username', repo: 'reponame', path: '.github/config.yml'}
   * ```
   *
   * @param object - Params to be merged with the repo params.
   *
   */
  public repo<T>(object?: T) {
    // @ts-ignore `repository` is not always present in this.payload
    const repo = this.payload.repository;

    if (!repo) {
      throw new Error(
        "context.repo() is not supported for this webhook event."
      );
    }

    return Object.assign(
      {
        owner: repo.owner.login || repo.owner.name,
        repo: repo.name,
      },
      object
    );
  }

  /**
   * Return the `owner`, `repo`, and `issue_number` params for making API requests
   * against an issue. The object passed in will be merged with the repo params.
   *
   *
   * ```js
   * const params = context.issue({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', issue_number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the issue params.
   */
  public issue<T>(object?: T) {
    return Object.assign(
      {
        issue_number:
          // @ts-ignore - this.payload may not have `issue` or `pull_request` keys
          (this.payload.issue || this.payload.pull_request || this.payload)
            .number,
      },
      this.repo(object)
    );
  }

  /**
   * Return the `owner`, `repo`, and `pull_number` params for making API requests
   * against a pull request. The object passed in will be merged with the repo params.
   *
   *
   * ```js
   * const params = context.pullRequest({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', pull_number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the pull request params.
   */
  public pullRequest<T>(object?: T) {
    const payload = this.payload;
    return Object.assign(
      {
        // @ts-ignore - this.payload may not have `issue` or `pull_request` keys
        pull_number: (payload.issue || payload.pull_request || payload).number,
      },
      this.repo(object)
    );
  }

  /**
   * Returns a boolean if the actor on the event was a bot.
   * @type {boolean}
   */
  get isBot() {
    // @ts-expect-error - `sender` key is currently not present in all events
    // see https://github.com/octokit/webhooks/issues/510
    return this.payload.sender.type === "Bot";
  }

  /**
   * Reads the app configuration from the given YAML file in the `.github`
   * directory of the repository.
   *
   * For example, given a file named `.github/config.yml`:
   *
   * ```yml
   * close: true
   * comment: Check the specs on the rotary girder.
   * ```
   *
   * Your app can read that file from the target repository:
   *
   * ```js
   * // Load config from .github/config.yml in the repository
   * const config = await context.config('config.yml')
   *
   * if (config.close) {
   *   context.octokit.issues.comment(context.issue({body: config.comment}))
   *   context.octokit.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * You can also use a `defaultConfig` object:
   *
   * ```js
   * // Load config from .github/config.yml in the repository and combine with default config
   * const config = await context.config('config.yml', {comment: 'Make sure to check all the specs.'})
   *
   * if (config.close) {
   *   context.octokit.issues.comment(context.issue({body: config.comment}));
   *   context.octokit.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * Config files can also specify a base that they extend. `deepMergeOptions` can be used
   * to configure how the target config, extended base, and default configs are merged.
   *
   * For security reasons, configuration is only loaded from the repository's default branch,
   * changes made in pull requests from different branches or forks are ignored.
   *
   * If you need more lower-level control over reading and merging configuration files,
   * you can `context.octokit.config.get(options)`, see https://github.com/probot/octokit-plugin-config.
   *
   * @param fileName - Name of the YAML file in the `.github` directory
   * @param defaultConfig - An object of default config options
   * @param deepMergeOptions - Controls merging configs (from the [deepmerge](https://github.com/TehShrike/deepmerge) module)
   * @return Configuration object read from the file
   */
  public async config<T>(
    fileName: string,
    defaultConfig?: T,
    deepMergeOptions?: MergeOptions
  ): Promise<T | null> {
    const params = this.repo({
      path: path.posix.join(".github", fileName),
      defaults(configs: object[]) {
        const result = merge.all(
          [defaultConfig || {}, ...configs],
          deepMergeOptions
        );

        return result;
      },
    });

    // @ts-ignore
    const { config, files } = await this.octokit.config.get(params);

    // if no default config is set, and no config files are found, return null
    if (!defaultConfig && !files.find((file: any) => file.config !== null)) {
      return null;
    }

    return config as T;
  }
}
