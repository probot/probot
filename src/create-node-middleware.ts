import { RequestListener } from "http";
import { createNodeMiddleware as createWebbhooksMiddleware } from "@octokit/webhooks";

import { ApplicationFunction } from "./types";
import { MiddlewareOptions } from "./types";

export function createNodeMiddleware(
  appFn: ApplicationFunction,
  { probot, webhooksPath }: MiddlewareOptions
): RequestListener {
  probot.load(appFn);

  return createWebbhooksMiddleware(probot.webhooks, {
    path: webhooksPath || "/",
  });
}
