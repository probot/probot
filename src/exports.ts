export type { Logger } from "pino";

export { Context } from "./context.js";

export { Probot } from "./probot.js";
export { Server } from "./server/server.js";
export { ProbotOctokit } from "./octokit/probot-octokit.js";
export { run } from "./run.js";
export { createNodeMiddleware } from "./create-node-middleware.js";
export { createProbot } from "./create-probot.js";

/** NOTE: exported types might change at any point in time */
export type {
  Options,
  ApplicationFunction,
  ApplicationFunctionOptions,
} from "./types.js";
