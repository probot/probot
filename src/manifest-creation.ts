import fs from "fs";
import yaml from "js-yaml";
import path from "path";
import updateDotenv from "update-dotenv";
import { ProbotOctokit } from "./octokit/probot-octokit";

export class ManifestCreation {
  get pkg() {
    let pkg: any;
    try {
      pkg = require(path.join(process.cwd(), "package.json"));
    } catch (e) {
      pkg = {};
    }
    return pkg;
  }

  public async createWebhookChannel() {
    try {
      // tslint:disable:no-var-requires
      const SmeeClient = require("smee-client");

      await this.updateEnv({
        WEBHOOK_PROXY_URL: await SmeeClient.createChannel(),
      });
    } catch (error) {
      // Smee is not available, so we'll just move on
      // tslint:disable:no-console
      console.warn("Unable to connect to smee.io, try restarting your server.");
    }
  }

  public getManifest(pkg: any, baseUrl: any) {
    let manifest: any = {};
    try {
      const file = fs.readFileSync(path.join(process.cwd(), "app.yml"), "utf8");
      manifest = yaml.safeLoad(file);
    } catch (error) {
      // App config does not exist, which is ok.
      if (error.code !== "ENOENT") {
        throw error;
      }
    }

    const generatedManifest = JSON.stringify(
      Object.assign(
        {
          description: manifest.description || pkg.description,
          hook_attributes: {
            url: process.env.WEBHOOK_PROXY_URL || `${baseUrl}/`,
          },
          name: process.env.PROJECT_DOMAIN || manifest.name || pkg.name,
          public: manifest.public || true,
          redirect_url: `${baseUrl}/probot/setup`,
          // TODO: add setup url
          // setup_url:`${baseUrl}/probot/success`,
          url: manifest.url || pkg.homepage || pkg.repository,
          version: "v1",
        },
        manifest
      )
    );

    return generatedManifest;
  }

  public async createAppFromCode(code: any) {
    const octokit = new ProbotOctokit();
    const options: any = {
      code,
      mediaType: {
        previews: ["fury"], // needed for GHES 2.20 and older
      },
      ...(process.env.GHE_HOST && {
        baseUrl: `${process.env.GHE_PROTOCOL || "https"}://${
          process.env.GHE_HOST
        }/api/v3`,
      }),
    };
    const response = await octokit.request(
      "POST /app-manifests/:code/conversions",
      options
    );

    const { id, client_id, client_secret, webhook_secret, pem } = response.data;
    await this.updateEnv({
      APP_ID: id.toString(),
      PRIVATE_KEY: `"${pem}"`,
      WEBHOOK_SECRET: webhook_secret,
      GITHUB_CLIENT_ID: client_id,
      GITHUB_CLIENT_SECRET: client_secret,
    });

    return response.data.html_url;
  }

  public async updateEnv(env: any) {
    // Needs to be public due to tests
    return updateDotenv(env);
  }

  get createAppUrl() {
    const githubHost = process.env.GHE_HOST || `github.com`;
    return `${
      process.env.GHE_PROTOCOL || "https"
    }://${githubHost}/settings/apps/new`;
  }
}
