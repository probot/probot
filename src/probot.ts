import LRUCache from "lru-cache";
import { Logger } from "pino";
import { EmitterWebhookEvent as WebhookEvent } from "@octokit/webhooks";

import { aliasLog } from "./helpers/alias-log";
import { auth } from "./auth";
import { getLog } from "./helpers/get-log";
import { getProbotOctokitWithDefaults } from "./octokit/get-probot-octokit-with-defaults";
import { getWebhooks } from "./octokit/get-webhooks";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { VERSION } from "./version";
import {
  ApplicationFunction,
  DeprecatedLogger,
  Options,
  ProbotWebhooks,
  State,
} from "./types";

export type Constructor<T> = new (...args: any[]) => T;

export class Probot {
  static version = VERSION;
  static defaults<S extends Constructor<any>>(this: S, defaults: Options) {
    const ProbotWithDefaults = class extends this {
      constructor(...args: any[]) {
        const options = args[0] || {};
        super(Object.assign({}, defaults, options));
      }
    };

    return ProbotWithDefaults;
  }

  public webhooks: ProbotWebhooks;
  public log: DeprecatedLogger;
  public version: String;
  public on: ProbotWebhooks["on"];
  public onAny: ProbotWebhooks["onAny"];
  public onError: ProbotWebhooks["onError"];
  public auth: (
    installationId?: number,
    log?: Logger
  ) => Promise<InstanceType<typeof ProbotOctokit>>;

  private state: State;

  constructor(options: Options = {}) {
    options.secret = options.secret || "development";

    let level = options.logLevel;
    const logMessageKey = options.logMessageKey;

    this.log = aliasLog(options.log || getLog({ level, logMessageKey }));

    // TODO: support redis backend for access token cache if `options.redisConfig`
    const cache = new LRUCache<number, string>({
      // cache max. 15000 tokens, that will use less than 10mb memory
      max: 15000,
      // Cache for 1 minute less than GitHub expiry
      maxAge: 1000 * 60 * 59,
    });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: Number(options.appId),
      privateKey: options.privateKey,
      cache,
      log: this.log,
      redisConfig: options.redisConfig,
      baseUrl: options.baseUrl,
    });
    const octokit = new Octokit();

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit,
      webhooks: {
        secret: options.secret,
      },
      appId: Number(options.appId),
      privateKey: options.privateKey,
      host: options.host,
      port: options.port,
    };

    this.auth = auth.bind(null, this.state);

    this.webhooks = getWebhooks(this.state);

    this.on = this.webhooks.on;
    this.onAny = this.webhooks.onAny;
    this.onError = this.webhooks.onError;

    this.version = VERSION;
  }

  public receive(event: WebhookEvent) {
    this.log.debug({ event }, "Webhook received");
    return this.webhooks.receive(event);
  }

  public async load(appFn: ApplicationFunction | ApplicationFunction[]) {
    if (Array.isArray(appFn)) {
      for (const fn of appFn) {
        await this.load(fn);
      }
      return;
    }

    return appFn(this, {});
  }
}
