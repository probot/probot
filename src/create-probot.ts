import { LogLevel, Options as PinoOptions } from "@probot/pino";
import { getPrivateKey } from "@probot/get-private-key";

import { getLog, GetLogOptions } from "./helpers/get-log";
import { Options } from "./types";
import { Probot } from "./probot";

type CreateProbotOptions = {
  overrides?: Options;
  defaults?: Options;
  env?: NodeJS.ProcessEnv;
};

const DEFAULTS = {
  APP_ID: "",
  WEBHOOK_SECRET: "",
  GHE_HOST: "",
  GHE_PROTOCOL: "",
  LOG_FORMAT: "",
  LOG_LEVEL: "warn",
  LOG_LEVEL_IN_STRING: "",
  LOG_MESSAGE_KEY: "msg",
  REDIS_URL: "",
  SENTRY_DSN: "",
};

/**
 * Merges configuration from defaults/environment variables/overrides and returns
 * a Probot instance. Finds private key using [`@probot/get-private-key`](https://github.com/probot/get-private-key).
 *
 * @see https://probot.github.io/docs/configuration/
 * @param defaults default Options, will be overwritten if according environment variable is set
 * @param overrides overwrites defaults and according environment variables
 * @param env defaults to process.env
 */
export function createProbot({
  overrides = {},
  defaults = {},
  env = process.env,
}: CreateProbotOptions = {}) {
  const privateKey = getPrivateKey({ env });
  const envWithDefaults = { ...DEFAULTS, ...env };

  const envOptions: Options = {
    logLevel: envWithDefaults.LOG_LEVEL as LogLevel,
    appId: Number(envWithDefaults.APP_ID),
    privateKey: (privateKey && privateKey.toString()) || undefined,
    secret: envWithDefaults.WEBHOOK_SECRET,
    redisConfig: envWithDefaults.REDIS_URL,
    baseUrl: envWithDefaults.GHE_HOST
      ? `${envWithDefaults.GHE_PROTOCOL || "https"}://${
          envWithDefaults.GHE_HOST
        }/api/v3`
      : "https://api.github.com",
  };

  const probotOptions = {
    ...defaults,
    ...envOptions,
    ...overrides,
  };

  const logOptions: GetLogOptions = {
    level: probotOptions.logLevel,
    logFormat: envWithDefaults.LOG_FORMAT as PinoOptions["logFormat"],
    logLevelInString: envWithDefaults.LOG_LEVEL_IN_STRING === "true",
    logMessageKey: envWithDefaults.LOG_MESSAGE_KEY,
    sentryDsn: envWithDefaults.SENTRY_DSN,
  };

  const log = getLog(logOptions).child({ name: "server" });

  return new Probot({
    log: log.child({ name: "probot" }),
    ...probotOptions,
  });
}
