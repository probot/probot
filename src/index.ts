import { Logger } from "pino";

import { Context } from "./context";
import { Options, ApplicationFunctionOptions } from "./types";
import { Probot } from "./probot";
import { Server } from "./server/server";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { run } from "./run";
import { createNodeMiddleware } from "./create-node-middleware";
import { createProbot } from "./create-probot";

export {
  Logger,
  Context,
  ProbotOctokit,
  run,
  Probot,
  Server,
  createNodeMiddleware,
  createProbot,
};

/** NOTE: exported types might change at any point in time */
export { Options, ApplicationFunctionOptions };
