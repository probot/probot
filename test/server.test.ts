import Stream from "stream";

import { NextFunction, Request, Response } from "express";
import request from "supertest";
import pino from "pino";
import { sign } from "@octokit/webhooks-methods";
import getPort from "get-port";

import { Server, Probot } from "../src";

const appId = 1;
const privateKey = `-----BEGIN RSA PRIVATE KEY-----
MIIBOQIBAAJBAIILhiN9IFpaE0pUXsesuuoaj6eeDiAqCiE49WB1tMB8ZMhC37kY
Fl52NUYbUxb7JEf6pH5H9vqw1Wp69u78XeUCAwEAAQJAb88urnaXiXdmnIK71tuo
/TyHBKt9I6Rhfzz0o9Gv7coL7a537FVDvV5UCARXHJMF41tKwj+zlt9EEUw7a1HY
wQIhAL4F/VHWSPHeTgXYf4EaX2OlpSOk/n7lsFtL/6bWRzRVAiEArzJs2vopJitv
A1yBjz3q2nX+zthk+GLXrJQkYOnIk1ECIHfeFV8TWm5gej1LxZquBTA5pINoqDVq
NKZSuZEHqGEFAiB6EDrxkovq8SYGhIQsJeqkTMO8n94xhMRZlFmIQDokEQIgAq5U
r1UQNnUExRh7ZT0kFbMfO9jKYZVlQdCL9Dn93vo=
-----END RSA PRIVATE KEY-----`;
const pushEvent = require("./fixtures/webhook/push.json");

describe("Server", () => {
  let server: Server;

  let output: any[];
  const streamLogsToOutput = new Stream.Writable({ objectMode: true });
  streamLogsToOutput._write = (object, encoding, done) => {
    output.push(JSON.parse(object));
    done();
  };

  beforeEach(async () => {
    output = [];
    const log = pino(streamLogsToOutput);
    server = new Server({
      Probot: Probot.defaults({
        appId,
        privateKey,
        secret: "secret",
        log: log.child({ name: "probot" }),
      }),
      log: log.child({ name: "server" }),
    });

    // Error handler to avoid printing logs
    server.expressApp.use(
      (error: Error, req: Request, res: Response, next: NextFunction) => {
        res.status(500).send(error.message);
      }
    );
  });

  test("Server.version", () => {
    expect(Server.version).toEqual("0.0.0-development");
  });

  describe("GET /ping", () => {
    it("returns a 200 response", async () => {
      await request(server.expressApp).get("/ping").expect(200, "PONG");
      expect(output.length).toEqual(1);
      expect(output[0].msg).toContain("GET /ping 200 -");
    });
  });

  describe("webhook handler (POST /)", () => {
    it("should return 200 and run event handlers in app function", async () => {
      expect.assertions(3);

      server = new Server({
        Probot: Probot.defaults({
          appId,
          privateKey,
          secret: "secret",
        }),
        log: pino(streamLogsToOutput),
        port: await getPort(),
      });

      await server.load((app) => {
        app.on("push", (event) => {
          expect(event.name).toEqual("push");
        });
      });

      const dataString = JSON.stringify(pushEvent);

      await request(server.expressApp)
        .post("/")
        .send(dataString)
        .set("x-github-event", "push")
        .set("x-hub-signature-256", await sign("secret", dataString))
        .set("x-github-delivery", "3sw4d5f6g7h8");

      expect(output.length).toEqual(1);
      expect(output[0].msg).toContain("POST / 200 -");
    });

    test("respond with a friendly error when x-hub-signature-256 is missing", async () => {
      await server.load(() => {});

      await request(server.expressApp)
        .post("/")
        .send(JSON.stringify(pushEvent))
        .set("x-github-event", "push")
        // Note: 'x-hub-signature-256' is missing
        .set("x-github-delivery", "3sw4d5f6g7h8")
        .expect(
          400,
          '{"error":"Required headers missing: x-hub-signature-256"}'
        );
    });
  });

  describe("GET unknown URL", () => {
    it("responds with 404", async () => {
      await request(server.expressApp).get("/notfound").expect(404);
      expect(output.length).toEqual(1);
      expect(output[0].msg).toContain("GET /notfound 404 -");
    });
  });

  describe(".start() / .stop()", () => {
    it("should expect the correct error if port already in use", (next) => {
      expect.assertions(1);

      // block port 3001
      const http = require("http");
      const blockade = http.createServer().listen(3001, async () => {
        const server = new Server({
          Probot: Probot.defaults({ appId, privateKey }),
          log: pino(streamLogsToOutput),
          port: 3001,
        });

        try {
          await server.start();
        } catch (error) {
          expect(error.message).toEqual(
            "Port 3001 is already in use. You can define the PORT environment variable to use a different port."
          );
        }

        await server.stop();
        blockade.close(() => next());
      });
    });

    it("should listen to port when not in use", async () => {
      const testApp = new Server({
        Probot: Probot.defaults({ appId, privateKey }),
        port: 3001,
        log: pino(streamLogsToOutput),
      });
      await testApp.start();

      expect(output.length).toEqual(2);
      expect(output[1].msg).toEqual("Listening on http://localhost:3001");

      await testApp.stop();
    });

    it("respects host/ip config when starting up HTTP server", async () => {
      const testApp = new Server({
        Probot: Probot.defaults({ appId, privateKey }),
        port: 3002,
        host: "127.0.0.1",
        log: pino(streamLogsToOutput),
      });
      await testApp.start();

      expect(output.length).toEqual(2);
      expect(output[1].msg).toEqual("Listening on http://127.0.0.1:3002");

      await testApp.stop();
    });
  });

  describe("router", () => {
    it("prefixes paths with route name", () => {
      const router = server.router("/my-app");
      router.get("/foo", (req, res) => res.end("foo"));

      return request(server.expressApp).get("/my-app/foo").expect(200, "foo");
    });

    it("allows routes with no path", () => {
      const router = server.router();
      router.get("/foo", (req, res) => res.end("foo"));

      return request(server.expressApp).get("/foo").expect(200, "foo");
    });

    it("allows you to overwrite the root path", () => {
      const router = server.router();
      router.get("/", (req, res) => res.end("foo"));

      return request(server.expressApp).get("/").expect(200, "foo");
    });

    it("isolates apps from affecting each other", async () => {
      ["foo", "bar"].forEach((name) => {
        const router = server.router("/" + name);

        router.use((req, res, next) => {
          res.append("X-Test", name);
          next();
        });

        router.get("/hello", (req, res) => res.end(name));
      });

      await request(server.expressApp)
        .get("/foo/hello")
        .expect(200, "foo")
        .expect("X-Test", "foo");

      await request(server.expressApp)
        .get("/bar/hello")
        .expect(200, "bar")
        .expect("X-Test", "bar");
    });

    it("responds with 500 on error", async () => {
      server.expressApp.get("/boom", () => {
        throw new Error("boom");
      });

      await request(server.expressApp).get("/boom").expect(500);
    });
  });
});
