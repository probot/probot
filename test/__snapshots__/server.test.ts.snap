// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Server > GET unknown URL > html formatted response > not-found-response 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /notfound</pre>
</body>
</html>
"
`;
